### PR TITLE
Fix #3273

### LINKED ISSUE #3273: String length validation error messages can't distinguish between singular and plural count.
Screenshot 2024-02-27 at 9 20 51 AM

It always uses the "(s)" universally either it is 1 or 0 or >1. 

For example, if the zod schema is like this:

```
const SignUpFormSchema = z.object({
    alphabet: z.string().min(1)
});
```

If the user tries to submit the form without inputting anything the error message will be shown as:

`String must contain at least 1 character(s)`

The error message can be improved to:

`String must contain at least 1 character`

### PR DESCRIPTION
**Pull request by sgrsngh2003**
String length validation error messages can't distinguish between singular and plural count.
Fix #3273

### CODE REVIEW

File: src/locales/en.ts [hunk 1]
```diff
@@ -71,10 +71,18 @@ const errorMap: ZodErrorMap = (issue, _ctx) => {
         message = `Array must contain ${
           issue.exact ? "exactly" : issue.inclusive ? `at least` : `more than`
         } ${issue.minimum} element(s)`;
-      else if (issue.type === "string")
-        message = `String must contain ${
+      else if (issue.type === "string"){
+        if (issue.minimum<=1){
```
Comment (m10rten) [inline src/locales/en.ts:75]: Add spaces between operators, variables and values.

`issue.minimum <= 1`

File: src/locales/en.ts [hunk 2]
```diff
@@ -71,10 +71,18 @@ const errorMap: ZodErrorMap = (issue, _ctx) => {
...
+          message = `String must contain ${
+            issue.exact ? "exactly" : issue.inclusive ? `at least` : `over`
+          } ${issue.minimum} character`
```
Comment (m10rten) [inline src/locales/en.ts:78]: You could just do:

```ts
message = `String must contain ${
            issue.exact ? "exactly" : issue.inclusive ? `at least` : `over`
          } ${issue.minimum} character${issue.minimum > 1 ? "(s)" : ""`;
```

**Review by m10rten — CHANGES_REQUESTED** [review]
Simple change does not need so much duplication

File: src/locales/en.ts (hunk 2 above)
Comment (sgrsngh2003) [reply src/locales/en.ts:78]: thanks for the explanation.

Comment (sgrsngh2003) [reply src/locales/en.ts:78]: should i do the changes or its done.

Comment (m10rten) [reply src/locales/en.ts:78]: Make the change, duplication is never nice.

Comment (sgrsngh2003) [reply src/locales/en.ts:78]: so i just overwrite my code with yours??

**Comment by colinhacks** [comment]
Thanks for the PR @sgrsngh2003, it's very well done. 

Unfortunately, a lot of Zod users have written tests that match against the exact error messages produced by Zod. While technically this could be merged as part of Zod 4, I'm not convinced the inconvenience of such a change for users is worth the marginal improvement to clarity here.

File: deno/lib/locales/en.ts [hunk 3]
```diff
@@ -71,10 +71,11 @@ const errorMap: ZodErrorMap = (issue, _ctx) => {
         message = `Array must contain ${
           issue.exact ? "exactly" : issue.inclusive ? `at least` : `more than`
         } ${issue.minimum} element(s)`;
-      else if (issue.type === "string")
+      else if (issue.type === "string"){
         message = `String must contain ${
           issue.exact ? "exactly" : issue.inclusive ? `at least` : `over`
-        } ${issue.minimum} character(s)`;
+          } ${issue.minimum} character${issue.minimum>1 ? "(s)" : ""}`;
```
Comment (santosmarco) [inline deno/lib/locales/en.ts:77]: Why are you keeping the `(` and `)` on `(s)` if now you know it must be plural?

Maybe:
```suggestion
          } ${issue.minimum} character${issue.minimum>1 ? "s" : ""}`;  
```
?

Also, in English, `0` quantities should also be pluralized: we say `0 apples`, not `0 apple`.

**Pull request closed without merging**